### ISSUE TITLE
PS2EXE-GUI (Write-Progress)

### ISSUE THREAD

**Issue by chrisrpry** [MScholtes/TechNet-Gallery#2]
the GUI does not seem to display child objects related to the parent object.
consider the following code which works in PS ISE/Console:

`($i1 = Get-Random (0..24));($i2 = Get-Random (25..49));($i3 = Get-Random (50..74))`
`do {`
`$TotalPc = [math]::Round(((@($i1;$i2;$i3)) | Measure-Object -Average).Average,0)`
`    Write-Progress -Id 999 -Activity "Parent" -Percentcomplete $TotalPc -Status "$($totalPc)%"`
`    Write-Progress -ParentId 999 -Id 0 -Activity "Child1" -Percentcomplete "$(if($i1 -lt 100){$i1++})$i1" -Status "$($i1)%"`
`    Write-Progress -ParentId 999 -Id 1 -Activity "Child2" -Percentcomplete "$(if($i2 -lt 100){$i2++})$i2" -Status "$($i2)%"`
`    Write-Progress -ParentId 999 -Id 2 -Activity "Child3" -Percentcomplete "$(if($i3 -lt 100){$i3++})$i3" -Status "$($i3)%"`
`    Start-Sleep -Milliseconds 500`
`}`
`until ($TotalPc -eq 100)`

it will show you the parent progress bar and below it, it's child progress bars.

can the tool be updated to include child progress bars similar to PS/ISE?

Thanks,

**Comment by MScholtes** [comment]
Hello chrisrpry,

implementing the commandlet **_Write-Progress_**  was very complex, so I decided not to support nesting (and also not to provide support for console mode). I hope you understand this, as I don't think it is a central feature.

Greetings

Markus

**Comment by chrisrpry** [comment]
Hi MScholtes,

Thanks for the prompt reply, I can appreciate how complex it can be! :-) as it was as complex to do natively.

I'm not certain what you mean with regards to console mode?
Currently I write my scripts, and then convert to GUI, and fully rely on the GUI representation of the progress bar.

as PS is just another script language, and short of "create-job" there isn't really much multi-tasking happening, whether console or GUI, there is the potential for a reasonable amount of human input/intervention for some scripts to continue.

where as at least with the Write-Progress, I can control within a loop, and change the information live on the screen.
which is very useful for end-users whom are running my scripts as published apps.
I cant speak for everyone's use case, though I do see write-progress as important as write-host or write-warning, write-error in PS2EXE-GUI.

as the above code was more or less sudo code, here is an excerpt of code from my script, as it may explain more clearly what i was trying to do.

```#set power action for each selected machine
Write-Progress -Id 999 -Activity "Requesting.."
foreach ($XYZMachine in $XYZCatalogMachines) {
#begin the power action requested
    If ($PwrAction -in "TurnOn","Resume","Restart","Reset") { #sets maintenance mode off
        $PwrMode = "On"
        Set-BrokerMachineMaintenanceMode -adminaddress $admAddr -InputObject $XYZMachine.MachineName $False  > $null
    }
    If ($PwrAction -in "Suspend", "Shutdown", "TurnOff") { # sets maintenance mode on
        $PwrMode = "Off"
        Set-BrokerMachineMaintenanceMode -adminaddress $admAddr -InputObject $XYZMachine.MachineName $True  > $null
    }
    New-BrokerHostingPowerAction -adminaddress $admAddr -Action $PwrAction -MachineName $XYZMachine.MachineName  > $null
}#end of foreach
#variables and hashtable globals for progress tracking
$startTime = (Get-Date)
$OveralProgress["Total"] = $XYZCatalogMachines.count # gets the total amount of machines selected
$OveralProgress["Percentage"] = 0
$OveralProgress["Completed"] = 0
$OveralProgress["Remaining"] = 0
$RefreshInterval1 = 999 #milliseconds
$RefreshInterval2 = 2500 #millisecons
$TimeSinceRefresh = $null # clear before script start.
#progress tracking
Do {
    #refresh interval 2
    If ($TimeSinceRefresh -gt $RefreshInterval2) {
        $TimeSinceRefresh = 0 # reset interval timer for this loop.
        foreach ($XYZMachine in $XYZCatalogMachines) {
            $XYZVmId = $($XYZCatalogMachines.IndexOf($XYZMachine)) #uses the index number in the array to give the machine a fixed ID during this loop
            #conditionally get data if power up or power down
            $XYZMachineState = ($XYZMachinePwr = get-brokermachine -adminaddress $admAddr -MachineName $XYZMachine.MachineName)
            $progressbarstatus = $(if ($XYZMachinePwr.PowerState -ne 'On') {$XYZMachinePwr.PowerState} else {$XYZMachineState.RegistrationState})
            Switch ($PwrMode) {
                On{if ($MachineProgress["XYZVm$XYZVmId"] -lt 100) {
                    switch($progressbarstatus){ off{$MachineProgress["XYZVm$XYZVmId"]++} TurningOn{if ($MachineProgress["XYZVm$XYZVmId"] -ge 33){$MachineProgress["XYZVm$XYZVmId"]++}else {$MachineProgress["XYZVm$XYZVmId"]=33}} Unregistered{if ($MachineProgress["XYZVm$XYZVmId"] -lt 50){$MachineProgress["XYZVm$XYZVmId"]=50};if (($MachineProgress["XYZVm$XYZVmId"] -ge 50) -and ($MachineProgress["XYZVm$XYZVmId"] -lt 97)){$MachineProgress["XYZVm$XYZVmId"]++}} initializing{$MachineProgress["XYZVm$XYZVmId"]=99} Registered{$MachineProgress["XYZVm$XYZVmId"]=100}}}
                    $filtermcs=@($XYZCatalogMachines.MachineName);$OveralProgress["Completed"] = (Get-Brokermachine -AdminAddress $admAddr -CatalogName $XYZMachineCatalogName.name -Filter {(RegistrationState -eq "Registered") -and (MachineName -in $filtermcs)}).count # updates hash table with total machines which are registered.
                }
                Off{if ($MachineProgress["XYZVm$XYZVmId"] -lt 100) {
                    switch($progressbarstatus){ Registered{$MachineProgress["XYZVm$XYZVmId"]++} Unregistered{if ($MachineProgress["XYZVm$XYZVmId"] -ge 33){$MachineProgress["XYZVm$XYZVmId"]++}else {$MachineProgress["XYZVm$XYZVmId"]=33}} TurningOff{if ($MachineProgress["XYZVm$XYZVmId"] -lt 50){$MachineProgress["XYZVm$XYZVmId"]=50};if (($MachineProgress["XYZVm$XYZVmId"] -ge 50) -and ($MachineProgress["XYZVm$XYZVmId"] -lt 97)){$MachineProgress["XYZVm$XYZVmId"]++}} off{$MachineProgress["XYZVm$XYZVmId"]=100}}}
                    $filtermcs=@($XYZCatalogMachines.MachineName);$OveralProgress["Completed"] = (Get-Brokermachine -AdminAddress $admAddr -CatalogName $XYZMachineCatalogName.name -Filter {(PowerState -eq "Off") -and (MachineName -in $filtermcs)}).count # updates hash table with total machines which are not on.
                }
            }
            #display progress bar for each machine, change % based on machine state
            $activity = "requested '$PwrAction' for '$($XYZMachine.MachineName)'"
            $status = "$($MachineProgress["XYZVm$XYZVmId"])% Completed"
            $percentComplete = $MachineProgress["XYZVm$XYZVmId"]
            $CurrentOperation = "machine status: $progressbarstatus"
            if (($condition = switch($PwrMode){ On{$progressbarstatus -ne "Registered"} Off{$progressbarstatus -ne "Off"}}) -eq $true) {
                Switch ($Host.name){ # put this here, as i'm not sure the GUI(EXE) supports child progress bars (seems to overlay them :( )
                    'Windows PowerShell ISE Host' {write-progress -ParentId 999 -Id $XYZVmId -activity $activity -Status $status -CurrentOperation $CurrentOperation -PercentComplete $percentComplete -ErrorAction SilentlyContinue > $null}
                    'ConsoleHost' {write-progress -ParentId 999 -Id $XYZVmId -activity $activity -Status $status -CurrentOperation $CurrentOperation -PercentComplete $percentComplete -ErrorAction SilentlyContinue > $null}
                    <#uncomment to enable child progress bars#># Default {write-progress -ParentId 999 -Id $XYZVmId -activity $activity -Status $status -CurrentOperation $CurrentOperation -PercentComplete $percentComplete -ErrorAction SilentlyContinue > $null}
                }  
            }
            else {
                write-progress -ParentId 999 -Id $XYZVmId -activity "'$PwrAction' for '$($XYZMachine.MachineName)' completed." -Completed -ErrorAction SilentlyContinue > $null;$MachineProgress["XYZVm$XYZVmId"] = 100
            }
            #log individual machine status to log file
            If ($progressbarstatus -ne $progressbarstatustemp) {Out-LogFile -InputObject "Machine Status '$($XYZMachine.MachineName)': $progressbarstatus" -Type INFO -DisableMsgs;$progressbarstatustemp = $progressbarstatus}
        }#end of foreach
    }#end of if
    $OveralProgress["Percentage"] = [math]::Round(($MachineProgress.values | Measure-Object -Average).Average,0) #updates the total average % of all and stores in hash table
    $OveralProgress["Remaining"] = $OveralProgress["Total"]-$OveralProgress["Completed"]
    #display parent progress bar
    $elapsed = ((Get-Date)-$startTime)
    $activity = "Requesting powerstate for machine(s) in catalog '$($XYZMachineCatalogName.name)', this can take up to 5 minutes (longer during busy periods)"
    $status = "Time elapsed: $($elapsed.minutes)m:$($elapsed.seconds)s ($($OveralProgress["Percentage"])% completed)"
    $percentComplete = $OveralProgress["Percentage"]
    $CurrentOperation = "$($OveralProgress["Remaining"]) VMs remaining ($($OveralProgress["Completed"]) of $($OveralProgress["Total"]) completed)"
    Write-Progress -Id 999 -Activity $activity -Status $status -PercentComplete $percentComplete -CurrentOperation $CurrentOperation
    #refresh interval 1
    Start-Sleep -Milliseconds $RefreshInterval1;$TimeSinceRefresh = $TimeSinceRefresh+$RefreshInterval1
}# end of main loop
Until (($OveralProgress["Percentage"] -eq 100) -and ($OveralProgress["Remaining"] -eq 0))```

again many thanks for the quick response!

**Comment by MScholtes** [comment]
Hello chrisrpry,

I will think about it....

Greetings
Markus

**Comment by chrisrpry** [comment]
Hi Markus,

Thank you considering it :)

Kind Regards
Chris

**Comment by MScholtes** [comment]
Hello Chris,

released the new version 0.5.0.21 of PS2EXE with support for nested progress bars today. Hope that is what you need.

Greetings

Markus

**Comment by chrisrpry** [comment]
Hello Markus,

Amazing! does exactly what I need, I hope this is useful for anyone else who needs this!
![InstructorPowerControl_Obscured](https://user-images.githubusercontent.com/67742634/87280702-5c690800-c4ea-11ea-8cb8-385896a634a3.jpg)

Thanks again for both your fast responses and solution 

Kind Regards, 

Chris.
